File: src/main/java/org/nexial/core/plugins/image/ImageCaptionHelper.java
/*
 * Copyright 2012-2018 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.nexial.core.plugins.image;

import java.awt.*;
import java.awt.geom.*;
import java.awt.image.*;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import javax.imageio.ImageIO;

import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.nexial.core.NexialConst.CommonColor;
import org.nexial.core.NexialConst.Image.ImageType;
import org.nexial.core.utils.ConsoleUtils;

import static java.awt.AlphaComposite.*;
import static java.awt.Font.*;
import static org.nexial.core.NexialConst.ImageCaption.*;
import static org.nexial.core.NexialConst.ImageCaption.CaptionPositions.BOTTOM_RIGHT;
import static org.nexial.core.NexialConst.replaceWindowsChars;

public class ImageCaptionHelper {

    public static class CaptionModel {
        private List<String> captions;
        private Color captionColor;
        private CaptionPositions position = BOTTOM_RIGHT;
        private String fontFace = DEF_FONT_FACE;
        private int fontSize = DEF_FONT_SIZE;
        private float alpha = DEF_ALPHA;

        public CaptionModel() {}

        public Color getCaptionColor() { return captionColor; }

        public void setCaptionColor(Color captionColor) { this.captionColor = captionColor; }

        public void setCaptionColor(String captionColor) { this.captionColor = CommonColor.toColor(captionColor); }

        public CaptionPositions getPosition() { return position; }

        public void setPosition(CaptionPositions position) {this.position = position;}

        public String getFontFace() {return fontFace;}

        public void setFontFace(String fontFace) {this.fontFace = fontFace;}

        public int getFontSize() {return fontSize;}

        public void setFontSize(int fontSize) {this.fontSize = fontSize;}

        public float getAlpha() {return alpha;}

        public void setAlpha(float alpha) {this.alpha = alpha;}

        public List<String> getCaptions() { return captions; }

        public void setCaptions(List<String> captions) { this.captions = captions; }

        public void addCaption(Object... captions) {
            if (ArrayUtils.isEmpty(captions)) { return; }
            if (this.captions == null) { this.captions = new ArrayList<>(); }
            Arrays.stream(captions).forEach(caption -> this.captions.add(StringUtils.trim(caption.toString())));
        }
    }

    public static void addCaptionToImage(File file, CaptionModel model) {

    // }
    //
    // public static void addCaptionToImage(File file, CaptionPositions position, Object... caption) {

        BufferedImage img;
        try {
            img = ImageIO.read(file);
        } catch (IOException e) {
            ConsoleUtils.error("Unable to read image " + file + ": " + e.getMessage());
            return;
        }

        if (img == null) { return; }

        int imageWidth = img.getWidth();
        int imageHeight = img.getHeight();
        if (imageWidth <= MIN_WIDTH || imageHeight <= MIN_HEIGHT) { return; }

        try {
            Graphics2D graphics = prepareGraphicObject(img, model);
            int lineHeight = graphics.getFontMetrics().getHeight();

            List<String> captions = new ArrayList<>();
            // Arrays.stream(caption).forEach(text -> captions.add(wrapCaptionName(imageWidth, (String) text, graphics)));
            model.getCaptions().forEach(text -> captions.add(wrapCaptionName(imageWidth, text, graphics)));

            List<String> captionList = getCaptionTextList(imageWidth, graphics, captions);

            if (captionList != null) {
                for (int line = 0; line < captionList.size(); line++) {
                    String captionString = captionList.get(line);
                    Rectangle2D rectangle;
                    rectangle = resolveBounds(graphics, captionString);
                    int lineCal = line * lineHeight;
                    // Pair<Integer, Integer> dimensions = getDimensions(img, position, rectangle, lineCal);
                    Pair<Integer, Integer> dimensions = getDimensions(img, model.position, rectangle, lineCal);
                    graphics.drawString(captionString, dimensions.getLeft(), dimensions.getRight());
                }
            }
            ImageIO.write(img, ImageType.png.toString(), file);
            graphics.dispose();

        } catch (Exception e) {
            ConsoleUtils.error("Unable to add caption to the image", e.getMessage());
        }
    }

    private static Pair<Integer, Integer> getDimensions(BufferedImage bufferedImage, CaptionPositions position,
                                                        Rectangle2D rectangle, int lineCal) {
        int captionWidth = 0;
        int captionHeight = 0;
        switch (position) {
            case BOTTOM_CENTER:
                captionWidth = (bufferedImage.getWidth() - (int) rectangle.getWidth()) / 2;
                captionHeight = (bufferedImage.getHeight() * paddingBottomHeight) / 100 + lineCal;
                break;
            case BOTTOM_LEFT:
                captionWidth = paddingLeftWidth;
                captionHeight = (bufferedImage.getHeight() * paddingBottomHeight) / 100 + lineCal;
                break;
            case BOTTOM_RIGHT:
                captionWidth = bufferedImage.getWidth() - (int) rectangle.getWidth();
                captionHeight = (bufferedImage.getHeight() * paddingBottomHeight) / 100 + lineCal;
                break;
            case MIDDLE_LEFT:
                captionWidth = paddingLeftWidth;
                captionHeight = (bufferedImage.getHeight() / 2) + lineCal;
                break;
            case MIDDLE_CENTER:
                captionWidth = (bufferedImage.getWidth() - (int) rectangle.getWidth()) / 2;
                captionHeight = (bufferedImage.getHeight() / 2) + lineCal;
                break;
            case MIDDLE_RIGHT:
                captionWidth = bufferedImage.getWidth() - (int) rectangle.getWidth();
                captionHeight = (bufferedImage.getHeight() / 2) + lineCal;
                break;
            case TOP_LEFT:
                captionWidth = paddingLeftWidth;
                captionHeight = paddingTopHeight + lineCal;
                break;
            case TOP_CENTER:
                captionWidth = (bufferedImage.getWidth() - (int) rectangle.getWidth()) / 2;
                captionHeight = paddingTopHeight + lineCal;
                break;
            case TOP_RIGHT:
                captionWidth = bufferedImage.getWidth() - (int) rectangle.getWidth();
                captionHeight = paddingTopHeight + lineCal;
                break;
            default:
                break;
        }
        return Pair.of(captionWidth, captionHeight);
    }

    private static List<String> getCaptionTextList(int imageWidth, Graphics2D graphics, List<String> captions) {

        if (captions.isEmpty()) { return null; }
        List<String> captionList = new ArrayList<>();
        int count = 0;
        String temp = "";

        while (count < captions.size()) {
            String combination;
            if (StringUtils.isNotEmpty(temp)) {
                combination = temp + " " + captions.get(count);
            } else {
                combination = temp + captions.get(count);
            }
            count = count + 1;
            if (checkCaptionFits(combination, imageWidth, graphics)) {
                temp = combination;
            } else {
                combination = captions.get(count - 1);
                if (!StringUtils.isEmpty(temp)) {captionList.add(temp);}
                temp = combination;
            }
        }

        if (!StringUtils.isEmpty(temp)) {captionList.add(temp);}
        return captionList;
    }

    private static boolean checkCaptionFits(String combination, int imageWidth, Graphics2D graphics) {
        Rectangle2D bounds = resolveBounds(graphics, combination);
        return imageWidth > bounds.getWidth();
    }

    private static Graphics2D prepareGraphicObject(BufferedImage img, CaptionModel model) {
        Graphics2D graphics = (Graphics2D) img.getGraphics();
        graphics.setComposite(AlphaComposite.getInstance(SRC_OVER, model.getAlpha()));
        graphics.setColor(model.getCaptionColor());
        graphics.setFont(new Font(model.getFontFace(), PLAIN, model.getFontSize()));
        return graphics;
    }

    private static String wrapCaptionName(int imageWidth, String caption, Graphics2D graphics) {
        Rectangle2D rect = resolveBounds(graphics, caption);

        // wrap the caption until caption rectangle width less than image width
        if (imageWidth > rect.getWidth()) { return caption; }

        while (rect.getWidth() > imageWidth) {
            caption = StringUtils.right(caption, caption.length() - trimIndex);
            rect = resolveBounds(graphics, caption);
        }
        return replaceWindowsChars.get("\\u2026") + caption;
    }

    private static Rectangle2D resolveBounds(Graphics2D graphics, String text) {
        return graphics.getFontMetrics().getStringBounds(text, graphics);
    }
}

File: src/main/kotlin/org/nexial/core/plugins/image/ImageCaptionHelper.kt
/*
 * Copyright 2012-2018 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.nexial.core.plugins.image

import org.apache.commons.lang3.StringUtils
import org.nexial.commons.utils.TextUtils
import org.nexial.core.NexialConst.CommonColor
import org.nexial.core.NexialConst.ImageCaption.*
import org.nexial.core.NexialConst.ImageCaption.CaptionPositions.*
import org.nexial.core.utils.ConsoleUtils
import java.awt.*
import java.awt.AlphaComposite.SRC_OVER
import java.awt.Font.PLAIN
import java.awt.Rectangle
import java.awt.font.FontRenderContext
import java.awt.font.LineBreakMeasurer
import java.awt.font.TextAttribute.*
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.text.AttributedCharacterIterator
import java.text.AttributedString
import javax.imageio.ImageIO
import kotlin.math.max
import kotlin.math.min

object ImageCaptionHelper {
    @JvmStatic
    fun addCaptionToImage(file: File, caption: CaptionModel) {

        val img: BufferedImage?
        try {
            img = ImageIO.read(file)
        } catch (e: IOException) {
            ConsoleUtils.error("Unable to read image " + file.toString() + ": " + e.message)
            return
        }

        if (img == null || img.width <= MIN_WIDTH || img.height <= MIN_HEIGHT) return

        val graphics = prepareGraphicObject(img, caption)
        try {
            renderText(graphics, resolveTextBounds(graphics, img, caption), caption)
            ImageIO.write(img, StringUtils.lowerCase(StringUtils.substringAfterLast(file.name, ".")), file)
        } catch (e: Exception) {
            ConsoleUtils.error("Unable to add caption to the image", e.message)
        } finally {
            // ConsoleUtils.log("Caption added to image " + file.getAbsolutePath());
            graphics.dispose()
        }
    }

    private fun resolveTextBounds(graphics: Graphics2D, img: BufferedImage, caption: CaptionModel): Rectangle {
        val imageWidth = img.width
        val imageHeight = img.height

        val maxTextBoundWidth = imageWidth.toDouble() / 4
        val maxTextBoundHeight = imageHeight.toDouble() / 4

        val textBound = graphics.fontMetrics.getStringBounds(caption.toText(), graphics)
        val oneLineHeight = textBound.height
        val oneLineWidth = textBound.width

        val width = min(maxTextBoundWidth, oneLineWidth).toInt()
        val height = (if (oneLineWidth > maxTextBoundWidth)
            min(maxTextBoundHeight, oneLineHeight * (oneLineWidth / maxTextBoundWidth))
        else
            oneLineHeight).toInt()

        val position = caption.position
        return Rectangle(when {
                             position.isLeft   -> LEFT_PADDING
                             position.isCenter -> (imageWidth - width) / 2
                             position.isRight  -> imageWidth - width - RIGHT_PADDING
                             else              -> 0
                         }, when {
                             position.isTop    -> TOP_PADDING
                             position.isMiddle -> (imageHeight - height) / 2
                             position.isBottom -> imageHeight - height - (LINE_PADDING * 2) - BOTTOM_PADDING
                             else              -> 0
                         }, width, height)
    }

    /**
     * render string on the specified [Graphics2D] instance. The `bounds` parameter specified the x, y,
     * width, height of the rendering space. The [CaptionModel] parameter represents the text and font setting
     * to render.
     */
    private fun renderText(g: Graphics2D, bounds: Rectangle, caption: CaptionModel): Rectangle {
        val text = caption.toText()
        if (StringUtils.isEmpty(text)) return Rectangle(bounds.x, bounds.y, 0, 0)

        val lineMeasurer = LineBreakMeasurer(createAttributedText(caption), FontRenderContext(null, true, true))

        val target = Point(bounds.x, bounds.y)
        var nextOffset: Int
        val align = caption.position

        if (align.isMiddle || align.isBottom) {
            if (align.isMiddle) target.y = bounds.y + bounds.height / 2
            if (align.isBottom) target.y = bounds.y + bounds.height - caption.fontSize

            while (lineMeasurer.position < text.length) {
                nextOffset = nextTextIndex(text, lineMeasurer.position, lineMeasurer.nextOffset(bounds.width.toFloat()))
                val textLayout = lineMeasurer.nextLayout(bounds.width.toFloat(), nextOffset, false)
                val totalTextHeight = textLayout.ascent + textLayout.leading + textLayout.descent
                if (align.isMiddle) target.y -= (totalTextHeight / 2).toInt()
                if (align.isBottom) target.y -= totalTextHeight.toInt()
            }
            target.y = max(TOP_PADDING, target.y)
            lineMeasurer.position = 0
        }

        if (align.isRight || align.isCenter) target.x = bounds.x + bounds.width

        val consumedBounds = Rectangle(target.x, target.y, 0, 0)
        while (lineMeasurer.position < text.length) {
            nextOffset = nextTextIndex(text, lineMeasurer.position, lineMeasurer.nextOffset(bounds.width.toFloat()))
            val textLayout = lineMeasurer.nextLayout(bounds.width.toFloat(), nextOffset, false)
            val textBounds = textLayout.bounds

            consumedBounds.width = max(consumedBounds.width, textBounds.width.toInt())

            target.y += textLayout.ascent.toInt() + LINE_PADDING

            when (align) {
                TOP_LEFT, MIDDLE_LEFT, BOTTOM_LEFT       -> {
                    textLayout.draw(g, target.x.toFloat(), target.y.toFloat())
                }

                TOP_CENTER, MIDDLE_CENTER, BOTTOM_CENTER -> {
                    target.x = bounds.x + bounds.width / 2 - (textBounds.width / 2).toInt()
                    consumedBounds.x = min(consumedBounds.x, target.x)
                    textLayout.draw(g, target.x.toFloat(), target.y.toFloat())
                }

                TOP_RIGHT, MIDDLE_RIGHT, BOTTOM_RIGHT    -> {
                    target.x = bounds.x + bounds.width - textBounds.width.toInt()
                    textLayout.draw(g, target.x.toFloat(), target.y.toFloat())
                    consumedBounds.x = max(consumedBounds.x, target.x)
                }
            }

            target.y += (textLayout.leading + textLayout.descent).toInt() + LINE_PADDING
        }

        consumedBounds.height = target.y - consumedBounds.y
        return consumedBounds
    }

    private fun prepareGraphicObject(img: BufferedImage, caption: CaptionModel): Graphics2D {
        val graphics = img.graphics as Graphics2D
        graphics.composite = AlphaComposite.getInstance(SRC_OVER, caption.alpha)
        // graphics.color = caption.getCaptionColor()
        graphics.font = Font(caption.fontFace, PLAIN, caption.fontSize)
        return graphics
    }

    private fun createAttributedText(caption: CaptionModel): AttributedCharacterIterator? {
        val attributedString = AttributedString(caption.toText())
        attributedString.addAttribute(FOREGROUND, caption.getCaptionColor())
        attributedString.addAttribute(FONT, Font(caption.fontFace, PLAIN, caption.fontSize))
        if (caption.withBackground) {
            attributedString.addAttribute(BACKGROUND,
                                          CommonColor.toComplementaryBackgroundColor(caption.getCaptionColor()))
        }
        attributedString.addAttribute(KERNING, LINE_PADDING)
        return attributedString.iterator
    }

    /** Calculates the next maximum index of the string that will be displayed.  */
    private fun nextTextIndex(text: String, from: Int, until: Int): Int {
        for (i in from + 1 until until) if (text[i] == '\n') return i
        return until
    }

    class CaptionModel {
        private var captions: MutableList<String>? = null
        private var captionColorText = "yellow"
        private var captionColor: Color? = CommonColor.toColor(captionColorText)
        var position = BOTTOM_RIGHT
        var alpha = DEF_ALPHA
        var fontFace = DEF_FONT_FACE
        var fontSize = DEF_FONT_SIZE
        var isWrap = false
        var withBackground = true

        fun getCaptionColor(): Color? {
            if (captionColor != null) return captionColor

            if (StringUtils.isEmpty(captionColorText)) {
                ConsoleUtils.log("No caption color specified; resorting to default")
                captionColorText = "yellow"
            }
            if (alpha < 0) {
                ConsoleUtils.log("No caption alpha setting found; resorting to default of 100%")
                alpha = 1f
            }
            captionColor = CommonColor.toColor(captionColorText)
            return captionColor
        }

        fun setCaptionColor(captionColor: String) {
            captionColorText = captionColor
            this.captionColor = CommonColor.toColor(captionColorText)
        }

        fun getCaptions(): List<String>? = captions

        fun setCaptions(captions: MutableList<String>?) {
            this.captions = captions
        }

        fun addCaptions(captions: List<Any>) {
            if (captions.isEmpty()) return
            if (this.captions == null) this.captions = mutableListOf()
            captions.forEach { this.captions!!.add(it.toString().trim()) }
        }

        fun toText(): String = TextUtils.toString(getCaptions(), if (isWrap) " " else "\n")
    }
}